classes/invitation/invitations/userRoleAssignment/UserRoleAssignmentInvite.php

```diff
@@ -0,0 +1,411 @@
+        InvitationModel::byStatus(InvitationStatus::PENDING)
+            ->byType(self::INVITATION_TYPE)
+            ->when(isset($this->invitationModel->userId), function ($query) {
+                return $query->byUserId($this->invitationModel->userId);
+            })
+            ->when(!isset($this->invitationModel->userId) && $this->invitationModel->email, function ($query) {
+                return $query->byEmail($this->invitationModel->email);
+            })
+            ->byContextId($this->invitationModel->contextId)
+            ->delete();
+    }
+
+    public function getInvitationActionRedirectController(): ?InvitationActionRedirectController
+    {
+        return new UserRoleAssignmentInviteRedirectController($this);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getCreateInvitationController(Invitation $invitation): CreateInvitationController 
+    {
+        return new UserRoleAssignmentCreateController($invitation);
+    }
+    
+    /**
+     * @inheritDoc
+     */
+    public function getReceiveInvitationController(Invitation $invitation): ReceiveInvitationController 
+    {
+        return new UserRoleAssignmentReceiveController($invitation);
+    }
+
+    public function getInviteValidationRules(): array
+    {
+        return ['customValidation' => [
+                function ($attribute, $value, $fail) {
+                    if (empty($this->invitation->userGroupsToAdd) && empty($this->invitation->userGroupsToRemove)) {
+                        $fail(__('invitation.userRoleAssignment.validation.error.noUserGroupChanges'));
+                    }
+                }
+            ]
+        ];
+    }
+
+    public function getFinaliseValidationRules(): array
+    {
+        return ['customValidation' => [
+                function ($attribute, $value, $fail) {
+                    $userId = $this->getUserId();
+
+                    if (isset($userId)) {
+                        $user = $this->getExistingUser();
+
+                        if (!isset($user)) {
+                            $fail(__('invitation.userRoleAssignment.validation.error.user.mustExist',
+                                [
+                                    'userId' => $userId
+                                ])
+                            );
+                        }
+                    }
+                    else if ($this->getEmail()) {
+                        $user = Repo::user()->getByEmail($this->getEmail());
+
+                        if (isset($user)) {
+                            $fail(__('invitation.userRoleAssignment.validation.error.user.emailMustNotExist', 
+                                [
+                                    'email' => $this->getEmail()
+                                ])
+                            );
+                        }
+                    }
+                }
+            ]
+        ];
+    }
+
+    public function getValidationRules(): array
```

Comment:
Rather than a function return a list of validation rules, why not a `function validate()` that applies them directly when called? (It's often a lot easier to work with immediate functions when there's not a reason for the indirection. Thinking e.g. of authorization rulesets, which are a pain to debug when they aren't working as expected.)

There's also a confusing list of functions here related to validation -- `validate` (which always returns `true`), `getValidationRules`, `getInviteValidationRules`, etc. We should have a single idiom for validation, and it should be clear what rules are applied when.